Upload ndbn csv

elasticspoon opened the pull request:
WIP

Currently errors look like this:

![image](https://github.com/rubyforgood/human-essentials/assets/14540596/0b6c5cc9-b43d-4fa6-b8b8-4a99bf33ff3a)

It is ugly otherwise it is hard to tell apart where one error starts and the next end. 
![image](https://github.com/rubyforgood/human-essentials/assets/14540596/2474f0a3-27f2-4a2a-8067-ac89a34874ab)


I could not figure out how to make them be on separate lines without changing the error partial.

veezus reviewed (commented):
Good work. That turned into a lot of work outside the core of uploading a CSV file!

elasticspoon commented on spec/system/ndbn_members_system_spec.rb:
```diff
@@ -0,0 +1,13 @@
+RSpec.describe "NDBN Members Upload", type: :system do
+  # TODO: This test crashes right after the upload
+  xit "CSV upload flow functions without error" do
+    user = create(:super_admin)
+    sign_in(user)
+    visit admin_ndbn_members_path
+
+    attach_file("member_file", "spec/fixtures/ndbn-large-import.csv")
+    click_button "Upload"
+
+    expect(page).to have_content("NDBN Members have been updated!")
+  end
+end
```
Not sure how to address this. Ideally I would like a system test just to run the full flow. In theory this should work but when you upload the file it gets attached and then the browser crashes.

elasticspoon replied:
Nice, it worked.

dorner commented on app/services/sync_ndbn_members.rb:
```diff
@@ -1,69 +1,45 @@
+  include ServiceObjectErrorsMixin
 
-  NDBN_MEMBERS_PAGE = "https://ndbn.memberclicks.net/member-ids".freeze
+  def initialize(member_file)
```
Is there a reason we need this to be an instance? Seems like all it's doing is calling a method and returning a set of error strings. We should be able to do that with a class-level method. Makes it much easier to test.

elasticspoon replied:
It mainly makes dealing with the errors easier. I have an object to attach the errors to, and then I can check if they exist in the controller.

The other option is having the service return either the parsed values or an error object? I could return something like `parsed_csv, error` and do a check like in Go. But I think using the `ServiceObjectErrorsMixin` makes sense and has precedence since its used in requests, etc.

dorner replied:
Yeah... I'm not the biggest fan of that mixin. :grin: In this case it doesn't need to return anything other than any possible errors. And... you know, they're errors. :grin: You can just raise an error from this and catch it on the calling side. That's really all that's needed here.

elasticspoon replied:
You know, I somehow totally blanked that the service was not actually returning anything. Its returning an array of errors now.

dorner commented on app/services/sync_ndbn_members.rb:
```diff
@@ -1,69 +1,45 @@
-          Rails.logger.info("New record found! #{ndbn_member.id} - #{ndbn_member.account_name} has been added!")
-        end
+      # Skip if nothing has changed!
+      if ndbn_member.persisted? && !ndbn_member.changed?
```
Not sure you need this - Rails is generally smart enough not to try to update a record if nothing has changed.

dorner commented on app/services/sync_ndbn_members.rb:
```diff
@@ -1,69 +1,45 @@
-      res.body
-    end
+  def parse_csv
+    return @_parsed if defined?(@_parsed)
```
Do we have a reason to think we'd be calling this method multiple times in a row?

dorner commented on app/views/admin/ndbn_members/index.html.erb:
```diff
@@ -0,0 +1,79 @@
+                <%- ndbn_members.each do |member| %>
+                  <tr>
+                    <td><%= member.ndbn_member_id %></td>
+                    <td><%= member.account_name %></td>
```
Do we need to paginate this? 

Actually, do we even need to display this data? We never have in the past. cc: @cielf

elasticspoon replied:
Just in case something's not working or you aren't sure that it uploaded. It could be useful just to be able to look at all the members. 

I think this was more useful before it became a super admin only feature. But it still might have some utility, I don't necessarily think you lose anything from it?

dorner replied:
Every piece of code we add comes with a hidden maintenance cost. E.g. when upgrading, it's another code path we need to ensure works. If we don't have a good business / use case for it, I'd prefer leaving it out. But I'll leave it up to @cielf who is likely going to be the resident super admin to decide. :grin:

elasticspoon replied:
@cielf sounds good. given we want to leave it in, do we want to paginate? I honestly feel like it might be better to not paginate just in case you want to quickly ctrl+f to search of a particular id or member.

cielf replied:
No need to paginate, IMO.

dorner commented on config/routes.rb:
```diff
@@ -71,6 +71,7 @@ def set_up_flipper
     end
     resources :questions
     resources :broadcast_announcements
+    resources :ndbn_members, only: %i(index create)
```
`create` doesn't match with REST semantics - it implies you're creating a single member, not doing a bulk upload. I'd use a different route name here (e.g. `upload_csv`).

dorner commented on spec/requests/ndbn_members_spec.rb:
```diff
@@ -0,0 +1,58 @@
+
+  describe "POST /create" do
+    it "updates the index contents" do
+      params = {member_file: fixture_file_upload("spec/fixtures/ndbn-large-import.csv", "text/csv")}
```
Why do we need to test a 90-line CSV file? A couple of lines should be enough.

elasticspoon replied:
Do you think the small CSV test is enough? I was doing it just to be thorough. I'm pretty sure the small CSV test has all the edge cases that are worth testing.

dorner replied:
Yep, I think the small one is fine.

elasticspoon commented:
@dorner Do you have any input on how to display the errors? I mention it in the PR description itself towards the end.

dorner commented:
You can do something like `errors.join("<br>".html_safe)` - something like that should work.

elasticspoon commented on app/views/shared/_flash.html.erb:
```diff
@@ -1,7 +1,8 @@
-    <div class="<%= flash_class(key) %> alert-dismissible" role="alert">
-      <%= message %>
-    </div>
+    <div
+      class="<%= flash_class(key) %> alert-dismissible"
+      role="alert"
+      style="white-space: pre-line;"><%= message %></div>
```
`errors.join("<br>".html_safe)` didn't work. It just renders at `stuff<br>stuff` instead of working like a real html element. Open to suggestions but this styling change makes the `\n` be respected.

cielf reviewed (changes requested):
I tried it out, and there are a couple of issues.
1/  I downloaded the file from the page,  changed the first member's id to xxxxx, and uploaded it.  It claimed to update, but that value did not change. 
2/  Could I get the list sorted either by name or by number, please?   Thanks.  Not a showstopper.

dorner commented:
Looks fine from a code perspective.

elasticspoon commented:
> 1/ I downloaded the file from the page, changed the first member's id to xxxxx, and uploaded it. It claimed to update, but that value did not change.

@cielf To clarify what are you expecting to happen? I have the file which is what seeds the data:

```csv
Updated: 1/10/2024,
NDBN Member Number,Member Name
10000,Homeless Shelter
20000,Other Spot
20000,
30000,Amazing Place
10000,Pawnee
```
```csv
Updated: 1/10/2024,
NDBN Member Number,Member Name
10000,Homeless Shelter
20000,Other Spot
20000,
30000,Amazing Place
10000,Pawnee
```
You upload:

```csv
Updated: 1/10/2024,
NDBN Member Number,Member Name
10000,Homeless Shelter
20000,Other Spot
20000,
30000,Amazing Place
10000,Pawnee
10000,Something
```

It should show Something, Homeless and Amazing.

The behavior is: if there is a matching id it updates the member name, otherwise makes a new member. It seems to be working that way on my end.
![image](https://github.com/rubyforgood/human-essentials/assets/14540596/1f7b8737-ca99-4d19-bf2f-c8982b3fe89f)


Also I messed up the seed file, push a fix since so if you re-seed get that.

elasticspoon commented:
Pushed a fix to seed file and to sort by id.

cielf commented:
So,  I downloaded the file from the NDBN page -- to make sure I had the right format.
Then I changed the first member id to xxxxxxxxxx
I would have expected to either get an error (if the member id is a flavour of numeric), or to have the member uploaded with the xxxxxxxxxx id (if it's text).     
What actually happened is that all the rest of the members uploaded, but that one didn't, and there was no error.

Here are the first few lines of that file, so you can see what I'm talking about:  
Updated: 1/10/2024,
NDBN Member Number,Member Name
xxxxxxxxxx,(914) Cares
12001,A Baby Center
22022,Access Period
21047,AIO Food & Energy Assistance

elasticspoon commented:
Gotcha. Hmmm. So the current behavior is that it will just silently drop items that don't fit the correct format which is: `some_numbers, Anything`. Would you prefer this some other way?

The reasoning was that the first line causes parsing issues. Frankly, I am not sure there is a case where there will be issues with the members being added but I have the errors there just in case.

cielf commented:
I'm not a fan of silent errors -- I don't think we have to rollback, though,  unless there is a way that we could make this error in a way that the properly formatted things would get messed up too.

cielf commented:
So, I guess what I'm saying is that I would like an error message if there is a typo in the id.   Might not ever be an issue,  but would be really hard to track down without that info.

elasticspoon commented:
Aite, new behavior:
- strip the first line if it contains "Update"
- attempt to parse the Member Number into a integer, if this fails add an error

This relies on two things, the headers not changing(this is optional, I could change it). And the assumption that the first line will be removed if it contains "Update".

dorner commented on app/services/sync_ndbn_members.rb:
```diff
@@ -6,26 +6,31 @@ def self.upload(member_file)
+        Integer(member_id)
+      rescue ArgumentError
+        ndbn_member.errors.add(:ndbn_member_id, "id must be an integer")
+      end
```
Probably could just check if `member_id.to_i.zero?`.

elasticspoon replied:
How about a regex match? Matching 0 would disallow "0" as an org number. Idk if that is a concern.

dorner replied:
I think we can disallow 0.

cielf replied:
There isn't going to be an ID of 0.  I'm pretty confident of that.

dorner commented on app/services/sync_ndbn_members.rb:
```diff
@@ -6,26 +6,31 @@ def self.upload(member_file)
+    data = member_file.readlines
+    data = data.drop(1) if /Update/.match?(data.first)
+    CSV.parse(data.join, headers: true, header_converters: :symbol)
+  rescue ArgumentError
```
Not sure I understand what's going on here?

elasticspoon replied:
basically, the CSV starts with `Updated: 23/132/123,` I need to remove that line to have the CSV parsing work. The second line is headers. 

The way I see it I have two options:
- strip the first line, parse the second as headers 
  - from what I can tell, I can't tell CSV to skip the first row. If I want to use the headers I need to change the CSV that is passed to the `CSV.parse` to not have that first row. So I open the file, drop the first row if its bad then pipe to CSV.
- strip both lines, use the data as an array of arrays

the second option is probably more clear so how about?
```rb
  def self.parse_csv(member_file)
    raise ParseError, "CSV upload is required." if member_file.nil?

    data = CSV.parse(member_file)
    # Remove the first row because csv starts with Updated: 1/10/2024,
    data = data.drop(1) if data.first&.first&.start_with?("Updated")
    # Remove the headers, we are not using them.
    data = data.drop(1) if data.first&.first&.match?(/NDBN/i)
  rescue ArgumentError
    raise ParseError, "The CSV File provided was invalid."
  end

```

dorner replied:
Hmm... not the biggest fan of this. Since it's only super users doing this, I'd rather force them to remove the first row manually before uploading. Then we can parse using CSV's `headers` parameter and everything "just works" a lot more flexibly.

We can add instructions to the page indicating that they need to ensure what the first and subsequent rows look like (by showing a sample).

@cielf thoughts?

cielf replied:
Well, I know I'm often somewhat distracted  -- what would the consequences be in this case if someone neglected to remove the line.

elasticspoon replied:
The headers would get messed up => every item gets parsed wrong => you get a 500 cookie overflow error because the session cookie runs out of space to show all the errors. 

The last part can be fixed by limiting the amount of errors shown but largely the result will be a bunch of errors shown like:

![image](https://github.com/rubyforgood/human-essentials/assets/14540596/9a73edcc-df9c-4225-b1fc-ca6b0dcc651a)

elasticspoon replied:
Could do something like 
![image](https://github.com/rubyforgood/human-essentials/assets/14540596/6ecf1710-f8b7-4f87-8d1e-aed7f42d161d)

dorner replied:
Can we not validate that the first line of the CSV looks the way we expect? Using something like this [https://dpericich.medium.com/how-to-use-ruby-to-get-csv-headers-3f3bffd3e965](https://dpericich.medium.com/how-to-use-ruby-to-get-csv-headers-3f3bffd3e965) to get the headers and immediately raise an error if they don't look the way we expect.

elasticspoon replied:
![image](https://github.com/rubyforgood/human-essentials/assets/14540596/f2e4c39a-f831-4ba2-9bf9-268b059bf311)

Man, this is why I love code review. We can certainly do that. I feel very stupid :upside_down_face:. Definitely tunneled a bit on my solution. Appreciate the more clear headed take.

cielf replied:
*nods* I think that will do for now -- in the longer term I'd still rather foist it off on NDBN, of course.

dorner commented on app/views/admin/ndbn_members/index.html.erb:
```diff
@@ -0,0 +1,79 @@
+          <div class="card-body">
+            <div class="row">
+              <p>Update the current NDBN members by uploading a CSV file of the
+                current members.</p>
```
Can we add some instructions here as to what format we expect?

elasticspoon replied:
@cielf any input on what the instructions should be? That fine?

![image](https://github.com/rubyforgood/human-essentials/assets/14540596/e868e7c1-8ca4-4536-bd1b-e07a75d647ff)

cielf replied:
We could give explicit instructions to say "The Current NDBN member list is found at https://docs.google.com/spreadsheets/d/1DRXeG206M0kgtUYT9lpD8fyEWRO7-a6m/edit?pli=1#gid=1068714915 .
Download the sheet, and remove the initial 'Updated' row.

If they change it again,  we can change the instructions as a Good First Issue.

elasticspoon replied:
![image](https://github.com/rubyforgood/human-essentials/assets/14540596/3c814dc3-bdcd-4630-977a-5f4d7fff5939)

New look

dorner commented on app/views/layouts/_lte_admin_sidebar.html.erb:
```diff
@@ -64,6 +64,12 @@
         <% end %>
       </li>
+      <li class="nav-item <%= active_class(['ndbm_members']) %>">
+        <%= link_to(admin_ndbn_members_path(organization_id: current_user.organization_id), class: "nav-link #{active_class(['ndbn_members'])}") do %>
```
You shouldn't need an organization\_id in this route, no?

dorner commented:
Looks good to me! @cielf did you want to take another look?
